runtime/src/snapshot_utils.rs

```diff
@@ -1769,6 +1803,102 @@ fn rebuild_bank_from_snapshots(
+}
+
+/// Verify that the snapshot's slot deltas are not corrupt/invalid
+/// These checks are simple/structural
+fn verify_slot_deltas_structural(
+    slot_deltas: &[BankSlotDelta],
+    bank_slot: Slot,
+) -> std::result::Result<(), VerifySlotDeltasError> {
+    // there should not be more entries than that status cache's max
+    let num_entries = slot_deltas.len();
+    if num_entries > status_cache::MAX_CACHE_ENTRIES {
+        return Err(VerifySlotDeltasError::TooManyEntries(
+            num_entries,
+            status_cache::MAX_CACHE_ENTRIES,
+        ));
+    }
+
+    let mut slots_seen_so_far = HashSet::new();
+    for &(slot, is_root, ..) in slot_deltas {
+        // all entries should be roots
+        if !is_root {
+            return Err(VerifySlotDeltasError::SlotIsNotRoot(slot));
+        }
+
+        // all entries should be for slots less than or equal to the bank's slot
+        if slot > bank_slot {
+            return Err(VerifySlotDeltasError::SlotGreaterThanMaxRoot(
+                slot, bank_slot,
+            ));
+        }
+
+        // there should only be one entry per slot
+        let inserted = slots_seen_so_far.insert(slot);
+        if !inserted {
+            return Err(VerifySlotDeltasError::SlotHasMultipleEntries(slot));
+        }
+    }
+
+    Ok(())
+}
+
+/// Verify that the snapshot's slot deltas are not corrupt/invalid
+/// These checks use the slot history for verification
+fn verify_slot_deltas_with_history(
+    slot_deltas: &[BankSlotDelta],
+    slot_history: &SlotHistory,
+    bank_slot: Slot,
+) -> std::result::Result<(), VerifySlotDeltasError> {
+    // all entries should be in the bank's slot history
+    let entry_missing_from_history = slot_deltas
+        .iter()
+        .find(|(slot, ..)| slot_history.check(*slot) != Check::Found);
+    if let Some((slot, ..)) = entry_missing_from_history {
+        return Err(VerifySlotDeltasError::SlotNotFoundInHistory(*slot));
+    }
+
+    // all slots in the history should be in the slot deltas (up to MAX_CACHE_ENTRIES)
+    // this ensures nothing was removed from the status cache
+
+    fn slot_deltas_contains_slot(slot_deltas: &[BankSlotDelta], slot: Slot) -> bool {
+        slot_deltas
+            .iter()
+            .any(|(slot_from_deltas, ..)| *slot_from_deltas == slot)
+    }
+
+    // go through the slot history and make sure there's an entry for each slot
+    // note: it's important to go highest-to-lowest since the status cache is a fifo
+    let mut entries_checked = 0;
+    for slot in (0..=bank_slot).rev() {
+        if entries_checked >= status_cache::MAX_CACHE_ENTRIES {
+            // We've gone through the whole slot deltas; done!
+            break;
+        }
+
+        if slot_history.check(slot) != Check::Found {
+            // This slot is not in the slot history; no need to check slot deltas
+            continue;
+        }
+
+        if !slot_deltas_contains_slot(slot_deltas, slot) {
```

Comment:
nit: i think this verification logic is of `O(N*N)` (i.e. quadratic) where N = MAX_CACHE_ENTRIES due to this repeated `fn` invocations inside this loop. I wonder whether this could be of concern or not, considering:

1. this is one-off code-path only used node setup.
2. the `N` is pretty fixed at the relatively low value, 300.

On the other hand, we have already a handy collection called `slots_seen_so_far`.


```rust
   // don't rename `slots_seen_so_far` in verify_slot_deltas_structural(); that's aptly named in the context
    ...

    let all_slots = verify_slot_deltas_structural(slot_deltas, bank.slot())?;
    assert_eq!(slot_deltas.len(), all_slots.len()); // detect serious logic error for future careless changes. :)
    verify_slot_deltas_with_history(slot_deltas, &all_slots, &bank.get_slot_history(), bank.slot())